provider/azure/instance_information.go

```diff
@@ -0,0 +1,33 @@
+package azure
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/juju/constraints"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/environs/instances"
+)
+
+var _ environs.InstanceTypesFetcher = (*azureEnviron)(nil)
+
+// InstanceTypes implements InstanceTypesFetcher
+func (env *azureEnviron) InstanceTypes(c constraints.Value) (instances.InstanceTypesWithCostMetadata, error) {
+	types, err := env.getInstanceTypes()
+	if err != nil {
+		return instances.InstanceTypesWithCostMetadata{}, errors.Trace(err)
+	}
+	result := make([]instances.InstanceType, len(types))
+	i := 0
+	for _, iType := range types {
+		result[i] = iType
+		i++
+	}
+	result, err = instances.MatchingInstanceTypes(result, "", c)
```

Comment:
I'm not *too* fussed, I'd just prefer not to pass the buck to the provider if we can avoid it, and also avoid repeating code in each provider. Maybe it's better to have the responsibility of the provider clear, and require that it honour the constraints.